### PR TITLE
[InstCombine] Fold `trunc nuw/nsw (x xor y) to i1` to `x != y`

### PR DESCRIPTION
**Pull request by YanWQ-monad**
Fold:
``` llvm
define i1 @src(i8 %x, i8 %y) {
  %xor = xor i8 %x, %y
  %r = trunc nuw i8 %xor to i1
  ret i1 %r
}

define i1 @tgt(i8 %x, i8 %y) {
  %r = icmp ne i8 %x, %y
  ret i1 %r
}
```

Proof: https://alive2.llvm.org/ce/z/dcuHmn

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 1]
```diff
@@ -770,6 +770,13 @@ Instruction *InstCombinerImpl::visitTrunc(TruncInst &Trunc) {
         return new ICmpInst(ICmpInst::Predicate::ICMP_EQ, X, Zero);
       }
     }
+
+    if (Trunc.hasNoUnsignedWrap()) {
+      Value *X, *Y;
+      if (match(Src, m_Xor(m_Value(X), m_Value(Y)))) {
+        return new ICmpInst(ICmpInst::ICMP_NE, X, Y);
+      }
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:778]:
```suggestion
      if (match(Src, m_Xor(m_Value(X), m_Value(Y))))
        return new ICmpInst(ICmpInst::ICMP_NE, X, Y);
```

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 2]
```diff
@@ -770,6 +770,12 @@ Instruction *InstCombinerImpl::visitTrunc(TruncInst &Trunc) {
         return new ICmpInst(ICmpInst::Predicate::ICMP_EQ, X, Zero);
       }
     }
+
+    if (Trunc.hasNoUnsignedWrap()) {
```
Comment (vfdff) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:774]: It also fine for nsw, https://alive2.llvm.org/ce/z/5fB_u9

**Review by vfdff — APPROVED** [review]
LGTM

**Comment by asmok-g** [comment]
Heads-up: we're seeing a probable mis-compile after this commit in a google internal test. The function that changes behavior with and without optimization is: https://github.com/google/arolla/blob/65c8389f271b9dce973801074217afb0e35f850c/arolla/memory/simple_buffer.h#L336.

Still checking if this is some UB in the code or a real mis-compile.

**Comment by asmok-g** [comment]
The "probable" mis-compile is only happening on arm.

**Comment by alexfh** [comment]
I suppose, this particular function is only involved in what _might_ be a miscompile, when it's inlined into another function, but disabling the optimization just on it using `#pragma clang optimize off/on` is enough to `fix` the test we see failing. The problem is that the whole setup leading to the problem is quite convoluted, and there's no isolated reproducer so far.

We're also actively investigating other possibilities like a preexisting UB in the code.

**Comment by alexfh** [comment]
False alarm, it seems. At least, we found UB in the code, fixing which resolves the problem. Apologies for the noise.

**Pull request merged**